apps/frontend/src/app/Solver/GOSolver/BNBSplitWorker/polyUB.ts

```diff
@@ -303,71 +319,89 @@ export type SumOfMonomials = Monomial[]
 type Monomial = {
   $k: number
   terms: string[]
+  setUsage: DynStat
 }
 function constM(v: number): Monomial {
-  return { $k: v, terms: [] }
+  return { $k: v, terms: [], setUsage: {} }
 }
-function weightedReadM(key: string, v: number): Monomial {
-  return { $k: v, terms: [key] }
+function termM(term: Term): Monomial {
+  const setUsage = term.artSet ? { [term.artSet.key]: term.artSet.thresh } : {}
+  return { $k: 1, terms: [term.key], setUsage }
 }
 function sumM(...monomials: Monomial[][]): Monomial[] {
   return monomials.flat()
 }
 function prodM(...monomials: Monomial[][]): Monomial[] {
-  return cartesian(...monomials).map((monos) =>
-    monos.reduce(
-      (ret, nxt) => {
-        ret.$k *= nxt.$k
-        ret.terms.push(...nxt.terms)
-        return ret
-      },
-      { $k: 1, terms: [] }
+  return cartesian(...monomials)
+    .map((monos) =>
+      monos.reduce(
+        (ret, nxt) => {
+          ret.$k *= nxt.$k
+          ret.terms.push(...nxt.terms)
+          Object.keys(nxt.setUsage).forEach(
+            (k) =>
+              (ret.setUsage[k] = Math.max(
+                ret.setUsage[k] ?? 0,
+                nxt.setUsage[k]
+              ))
+          )
+          return ret
+        },
+        { $k: 1, terms: [], setUsage: {} }
+      )
+    )
+    .filter(
+      ({ setUsage }) => Object.values(setUsage).reduce((a, b) => a + b, 0) < 5
```

Comment:
We have 5 artifact slots, so shouldn't be
```suggestion
      ({ setUsage }) => Object.values(setUsage).reduce((a, b) => a + b, 0) <= 5
```